File: shell/plasma/package/contents/ui/logic/Account.js
.pragma library

.import "DaemonText.js" as DaemonText
.import "FormatTime.js" as FormatTime
.import "I18n.js" as I18n
.import "Recovery.js" as Recovery
.import "State.js" as State

const SIGNED_OUT_ERRORS = ["not_signed_in", "sign_in_expired"];
const SPEND_PERIODS = [[I18n.N("Today"), "today"], [I18n.N("Yesterday"), "yesterday"], [I18n.N("Last 30 Days"), "last30Days"]];

function failedOffline(account, offline) {
    return offline && account.error?.kind === "network";
}

function statusSlot(account, offline) {
    if (account.status === "refreshing")
        return "refreshing";
    if (account.status === "stale" || (account.status === "error" && failedOffline(account, offline)))
        return "outdated";
    if (account.status === "error")
        return "warning";
    return "";
}

function isQuietlyLimited(account) {
    return account.error?.kind === "rate_limited" && (account.updatedAt ?? null) !== null;
}

function rateLimitNote(lang, account, timeFormat) {
    if (!isQuietlyLimited(account))
        return "";
    const next = account.refresh?.nextAt ?? null;
    if (next === null)
        return I18n.tr(lang, "Provider is limiting requests");
    return I18n.tr(lang, "Provider is limiting requests · next try {time}", {
        time: FormatTime.clockTime(next, timeFormat)
    });
}

function settledStatus(account) {
    if (account.status !== "refreshing" || account.error === null || isQuietlyLimited(account))
        return account.status;
    if (SIGNED_OUT_ERRORS.includes(account.error.kind))
        return "signed_out";
    if (account.error.kind === "no_subscription")
        return "no_subscription";
    return "error";
}

function isSignedOut(account) {
    return settledStatus(account) === "signed_out";
}

function signedOutNotice(lang, account, providers) {
    return {
        kind: "signin",
        title: I18n.tr(lang, "Signed out of {provider}", {
            provider: account.providerName
        }),
        detail: Recovery.signedOutDetail(lang, account),
        note: "",
        actions: Recovery.actions(lang, account, providers, true)
    };
}

function daemonNote(account) {
    const error = account.error;
    return error === null || error.message === error.kind ? "" : error.message;
}

function noSubscriptionNotice(lang, account) {
    return {
        kind: "warning",
        title: I18n.tr(lang, "No active subscription"),
        detail: I18n.tr(lang, "Limits aren't available for this account. Renew the plan or sign in with another account."),
        note: daemonNote(account),
        actions: [Recovery.retryAction(lang, account)]
    };
}

function errorNotice(lang, account, providers) {
    return {
        kind: "error",
        title: I18n.tr(lang, "Couldn't refresh {provider}", {
            provider: account.providerName
        }),
        detail: account.error?.message ?? "",
        note: Recovery.terminalHint(lang, account),
        actions: Recovery.actions(lang, account, providers, false)
    };
}

function providerKind(tone) {
    if (tone === "critical")
        return "error";
    if (tone === "warning")
        return "warning";
    return "info";
}

function providerNotice(lang, notice) {
    return {
        kind: providerKind(notice.tone),
        title: DaemonText.notice(lang, notice.text),
        detail: "",
        note: "",
        actions: []
    };
}

function notices(lang, account, offline, providers) {
    if (isSignedOut(account))
        return [signedOutNotice(lang, account, providers)];
    const status = settledStatus(account);
    if (status === "no_subscription")
        return [noSubscriptionNotice(lang, account)];
    const rows = account.notices.map(notice => providerNotice(lang, notice));
    if (status === "error" && !failedOffline(account, offline))
        rows.unshift(errorNotice(lang, account, providers));
    return rows;
}

function plates(notices) {
    return notices.filter(notice => notice.kind !== "info");
}

function infoLines(notices) {
    return notices.filter(notice => notice.kind === "info");
}

function showsQuotas(account) {
    return State.hasQuotas(account) && !isSignedOut(account);
}

function showsSpend(account, display) {
    return account.usage !== null && display.showAccountSpend;
}

function showsTrend(account, display) {
    return showsQuotas(account) && account.usage !== null && display.showTrend;
}

function hasExtras(account, display) {
    return showsQuotas(account) && (showsSpend(account, display) || account.balances.length > 0);
}

function extrasAlwaysOpen(account) {
    return showsQuotas(account) && State.shownWindows(account).length === 0 && account.balances.length > 0;
}

function spendRows(lang, account, display) {
    if (!showsSpend(account, display))
        return [];
    const provider = account.providerName;
    return SPEND_PERIODS.map(([msgid, key]) => {
        const title = I18n.tr(lang, msgid);
        return {
            title,
            breakdownTitle: `${title} · ${provider}`,
            totals: account.usage[key]
        };
    });
}

function removal(lang, account) {
    if (account.owner === "headroom")
        return {
            subtitle: I18n.tr(lang, "Deletes the sign-in Headroom created for this account"),
            confirmation: I18n.tr(lang, "Headroom deletes the sign-in it created for this account. The account itself is not affected.")
        };
    return {
        subtitle: I18n.tr(lang, "Stops showing this account; the {provider} CLI stays signed in", {
            provider: account.providerName
        }),
        confirmation: I18n.tr(lang, "Headroom will stop showing this account. The {provider} CLI stays signed in; you can sign in again through Headroom.", {
            provider: account.providerName
        })
    };
}

File: crates/headroom-tray/src/account.rs
use crate::i18n::{Lang, fill};
use crate::notices::{NoticeKind, notice_text};
use crate::payload::{Account, AccountError, Status, Window};
use crate::recovery::{NoticeButton, cli_login_hint, notice_buttons};

const NO_SUBSCRIPTION: &str = "no_subscription";
const NETWORK: &str = "network";
const ACCOUNT_CHANGED: &str = "account_changed";
const SIGN_IN_ERRORS: [&str; 2] = ["not_signed_in", "sign_in_expired"];
const SKELETON_ROWS: usize = 2;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HeaderStatus {
    Refreshing,
    Outdated,
    Error,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NoticeView {
    pub kind: NoticeKind,
    pub title: String,
    pub detail: Option<String>,
    pub note: Option<String>,
    pub buttons: Vec<NoticeButton>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum CardBody<'a> {
    Blocked(NoticeView),
    Skeleton(usize),
    Limits {
        alert: Option<NoticeView>,
        notices: Vec<NoticeView>,
        windows: Vec<&'a Window>,
    },
}

fn error_kind(account: &Account) -> Option<&str> {
    account.error.as_ref().map(|error| error.kind.as_str())
}

#[must_use]
pub fn lacks_subscription(account: &Account) -> bool {
    account.status == Status::NoSubscription
        || (account.status == Status::Refreshing && error_kind(account) == Some(NO_SUBSCRIPTION))
}

#[must_use]
pub fn is_signed_out(account: &Account) -> bool {
    account.status == Status::SignedOut
        || (account.status == Status::Refreshing
            && error_kind(account).is_some_and(|kind| SIGN_IN_ERRORS.contains(&kind)))
}

#[must_use]
pub fn is_retrying(account: &Account) -> bool {
    account.status == Status::Refreshing
}

fn failed_offline(account: &Account, offline: bool) -> bool {
    offline && error_kind(account) == Some(NETWORK)
}

#[must_use]
pub fn header_status(account: &Account, offline: bool) -> Option<HeaderStatus> {
    match account.status {
        Status::Refreshing => Some(HeaderStatus::Refreshing),
        Status::Stale => Some(HeaderStatus::Outdated),
        Status::Error if failed_offline(account, offline) => Some(HeaderStatus::Outdated),
        Status::Error => Some(HeaderStatus::Error),
        _ => None,
    }
}

#[must_use]
pub fn shows_name(account: &Account, accounts: &[&Account]) -> bool {
    accounts
        .iter()
        .filter(|other| other.provider == account.provider)
        .count()
        > 1
}

#[must_use]
pub fn account_title(account: &Account, show_name: bool) -> String {
    let who = account.label.as_ref().or(account.email.as_ref());
    match who {
        Some(who) if show_name => format!("{}: {who}", account.provider_name),
        _ => account.provider_name.clone(),
    }
}

#[must_use]
pub fn shown_plan(account: &Account) -> Option<&str> {
    account
        .plan
        .as_deref()
        .filter(|_| !lacks_subscription(account))
}

fn normalized(text: &str) -> String {
    let spaced: String = text
        .to_lowercase()
        .chars()
        .map(|c| if c.is_alphanumeric() { c } else { ' ' })
        .collect();
    spaced.split_whitespace().collect::<Vec<_>>().join(" ")
}

#[must_use]
pub fn subscription_note(error: Option<&AccountError>) -> Option<String> {
    let error = error?;
    let text = normalized(&error.message);
    let redundant = error.message == error.kind
        || text.is_empty()
        || text == "no active subscription"
        || text == normalized(NO_SUBSCRIPTION);
    (!redundant).then(|| error.message.clone())
}

fn signed_out_notice(lang: Lang, account: &Account, terminal_sign_in: bool) -> NoticeView {
    let detail = cli_login_hint(lang, account).unwrap_or_else(|| {
        lang.tr("Sign in again to keep this account up to date.")
            .to_owned()
    });
    NoticeView {
        kind: NoticeKind::SignIn,
        title: fill(
            lang.tr("Signed out of {provider}"),
            &[("provider", &account.provider_name)],
        ),
        detail: Some(detail),
        note: account.error.as_ref().map(|error| error.message.clone()),
        buttons: notice_buttons(account, true, terminal_sign_in),
    }
}

fn no_subscription_notice(lang: Lang, account: &Account) -> NoticeView {
    NoticeView {
        kind: NoticeKind::Warning,
        title: lang.tr("No active subscription").to_owned(),
        detail: Some(
            lang.tr(
                "Limits aren't available for this account. Renew the plan or sign in with another account.",
            )
            .to_owned(),
        ),
        note: subscription_note(account.error.as_ref()),
        buttons: notice_buttons(account, false, false),
    }
}

fn error_title(lang: Lang, account: &Account) -> String {
    let template = if error_kind(account) == Some(ACCOUNT_CHANGED) {
        "Another account is signed in to {provider}"
    } else {
        "Couldn't refresh {provider}"
    };
    fill(lang.tr(template), &[("provider", &account.provider_name)])
}

fn error_notice(lang: Lang, account: &Account) -> NoticeView {
    NoticeView {
        kind: NoticeKind::Error,
        title: error_title(lang, account),
        detail: account.error.as_ref().map(|error| error.message.clone()),
        note: cli_login_hint(lang, account),
        buttons: notice_buttons(account, false, false),
    }
}

fn daemon_notices(lang: Lang, account: &Account) -> impl Iterator<Item = NoticeView> + '_ {
    account.notices.iter().map(move |notice| NoticeView {
        kind: NoticeKind::from_tone(notice.tone),
        title: notice_text(lang, &notice.text),
        detail: None,
        note: None,
        buttons: Vec::new(),
    })
}

#[must_use]
pub fn shown_windows(account: &Account) -> Vec<&Window> {
    account
        .windows
        .iter()
        .filter(|window| !window.hidden)
        .collect()
}

fn awaiting_first_data(account: &Account) -> bool {
    account.status == Status::Refreshing && account.updated_at.is_none()
}

#[must_use]
pub fn shows_error_notice(account: &Account, offline: bool) -> bool {
    let reported = match account.status {
        Status::Error => true,
        Status::Refreshing => account.error.is_some(),
        _ => false,
    };
    reported && !failed_offline(account, offline)
}

#[must_use]
pub fn card_body(
    lang: Lang,
    account: &Account,
    offline: bool,
    terminal_sign_in: bool,
) -> CardBody<'_> {
    if is_signed_out(account) {
        return CardBody::Blocked(signed_out_notice(lang, account, terminal_sign_in));
    }
    if lacks_subscription(account) {
        return CardBody::Blocked(no_subscription_notice(lang, account));
    }
    let failed = shows_error_notice(account, offline);
    if awaiting_first_data(account) && !failed {
        return CardBody::Skeleton(SKELETON_ROWS.max(account.windows.len()));
    }
    CardBody::Limits {
        alert: failed.then(|| error_notice(lang, account)),
        notices: daemon_notices(lang, account).collect(),
        windows: shown_windows(account),
    }
}

#[cfg(test)]
#[path = "account_tests.rs"]
mod tests;
